02-react-week-1/state.md

```diff
@@ -0,0 +1,291 @@
+      </button>
+    </div>
+
+    );
+    }
+}
+
+export default Student;
+```
+
+Now, lets see what it looks like as a functional component  using the `useState` hook.  First, let's roll back to the functional component version we had before but add `useState` at the top:
+
+```javascript
+import React, { useState } from 'react';
+
+const Student = (props) => {
+  return (
+    <div>
+      <h3>{props.fullName}</h3>
+      <ul>
+        <li>Class: C14</li>
+        <li>Birthday: {props.birthday}</li>
+        <li>Email: {props.email}</li>
+      </ul>
+    </div>
+  );
+};
+
+```
+
+Next, we'll set up our state variable:
+
+```javascript
+import React, { useState } from 'react';
+
+const Student = (props) => {
+  
+  const [present, setPresent] = useState(false);
+
+  return (
+    // ...
+  );
+};
+
+```
+
+Last, we'll add our button and the function `changePresent`:
+
+```javascript
+import React, { useState } from 'react';
+
+const Student = (props) => {
+  
+  const [present, setPresent] = useState(false);
+
+  const changePresent = () => setPresent(!present);
+
+  // Component functions always return JSX
+  return (
+    <div>
+      <h3>{props.fullName}</h3>
+      <ul>
+        <li>Class: C13</li>
+        <li>Birthday: {props.birthday}</li>
+        <li>Email: {props.email}</li>
+      </ul>
+      // Note that we will get into what's happening in this line in the Events lesson, coming up next!
+      <button onClick={changePresent}>
+        Mark {present ? 'Absent' : 'Present'}
+      </button>
+    </div>
+  );
+};
+
+export default Student;
+
+```
+
+And we're done!  The functional component version is shorter, syntactically simpler to write and easier to read.  Since the addition of hooks, functional components have become the preferred way of writing components.
+
+
+## PacMan
+
+So far all of our examples have been fairly simple.  Let's look at how state can be used in a more complex way to build the game (PacMan)[https://en.wikipedia.org/wiki/Pac-Man]!  The elements of PacMan are PacMan himself, the ghosts, the dots, the special energizer dots and the fruit.  We could store all of the data at the App level and propagate it down using props.  If we look at all of the moving parts in PacMan, this solution quickly becomes very intricate and requires a lot of logic at the App level. If we think about this problem in terms of state, and letting each element own it's own state, things become much simpler.  Let's start from assuming we have set up a callback structure that allows our elements to know where PacMan is, and that they each have internal logic about what to do if they intersect with PacMan.  We're not going to cover that part of the project in this lesson, we're just going to look at how we could use state to keep track of important information inside each component.  Take a look at this diagram and think about how the state of each component would change as the game progresses:
```

Comment:
The only criticism I have with Pacman here is that because the sibling components here would have to interact, you'd have to "lift" state for the game to function in a lot of these examples, so it's a bit misleading.  Maybe if we have a later lesson on lifting state.